python/cugraph_service/cugraph_service_client/remote_graph_utils.py

```diff
@@ -0,0 +1,184 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import importlib
+import numpy as np
+
+
+class MissingModule:
+    """
+    Raises RuntimeError when any attribute is accessed on instances of this
+    class.
+
+    Instances of this class are returned by import_optional() when a module
+    cannot be found, which allows for code to import optional dependencies, and
+    have only the code paths that use the module affected.
+    """
+
+    def __init__(self, mod_name):
+        self.name = mod_name
+
+    def __getattr__(self, attr):
+        raise RuntimeError(f"This feature requires the {self.name} " "package/module")
+
+
+try:
+    cudf = importlib.import_module("cudf")
+except ModuleNotFoundError:
+    cudf = MissingModule("cudf")
+
+try:
+    cupy = importlib.import_module("cupy")
+except ModuleNotFoundError:
+    cupy = MissingModule("cupy")
+
+try:
+    pandas = importlib.import_module("pandas")
+except ModuleNotFoundError:
+    pandas = MissingModule("pandas")
+
+try:
+    torch = importlib.import_module("torch")
+except ModuleNotFoundError:
+    torch = MissingModule("torch")
+
+
+def _transform_to_backend_dtype_1d(data, series_name=None, backend="numpy", dtype=None):
+    """
+    Supports method-by-method selection of backend type (cupy, cudf, etc.)
+    to avoid costly conversion such as row-major to column-major transformation.
+    This method is used for 1-dimensional data, and does not perform unncessary
+    transpositions or copies.
+
+    Note: If using inferred dtypes, the returned series, array, or tensor may
+    infer a different dtype than what was originally on the server (i.e promotion
+    of int32 to int64).  In the future, the server may also return dtype to prevent
+    this from occurring.
+
+    data : np.ndarray
+        The raw ndarray that will be transformed to the backend dtype.
+    series_name : string
+        The name of the series (only used for dataframe backends).
+    backend : ('numpy', 'pandas', 'cupy', 'cudf', 'torch', 'torch:<device>')
+              [default = 'numpy']
+    dtype : ('int32', 'int64', 'float32', etc.)
+        Optional. The data type to use when storing data in a series or array.
+        If not set, it will be inferred for dataframe backends, and assumed as float64
+        for array and tensor backends.
+
+    """
+
+    if backend == "numpy":
+        return np.array(data, dtype=dtype or "float64")
```

Comment:
in the case of input numpy to numpy backend, is this only useful if the caller wants a different dtype?  If so, can you return a [view ](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.view.html) instead, or is a copy always required?

edit: according to the docstring for this function, it seems like a copy is not desired so maybe a view would be best afterall?

Reply:
View won't work, but I will add a way to prevent the copy if the dtype matches.  See below.

```
>>> import numpy as np
>>> a = np.array([0,1,2,3,4,5])
>>> a
array([0, 1, 2, 3, 4, 5])
>>> a.dtype
dtype('int64')
>>> a.view(dtype='int8')
array([0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0,
       0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0,
       0, 0, 0, 0], dtype=int8)
```